File: code_samples/section5/problems/problem5_1/problem5_1.cs
#nullable enable  // Enable nullable reference type analysis (TreeNode? means it may be null)

// ==========================
// INORDER TRAVERSAL (DFS)
// ==========================

// Returns the inorder traversal of a binary tree as a list of ints
// Order: Left -> Root -> Right
IList<int> Inorder(TreeNode? root)
{
    // This list will collect the traversal output
    var result = new List<int>();

    // Local recursive DFS function that performs the traversal
    void Dfs(TreeNode? node)
    {
        // Base case: empty subtree
        if (node == null) return;

        // Visit left subtree first
        Dfs(node.Left);

        // Visit current node
        result.Add(node.Val);

        // Visit right subtree
        Dfs(node.Right);
    }

    // Start DFS from the root
    Dfs(root);

    // Return the collected results
    return result;
}

// ==========================
// POSTORDER TRAVERSAL (DFS)
// ==========================

// Returns the postorder traversal of a binary tree as a list of ints
// Order: Left -> Right -> Root
IList<int> Postorder(TreeNode? root)
{
    // This list will collect the traversal output
    var result = new List<int>();

    // Local recursive DFS function that performs the traversal
    void Dfs(TreeNode? node)
    {
        // Base case: empty subtree
        if (node == null) return;

        // Visit left subtree
        Dfs(node.Left);

        // Visit right subtree
        Dfs(node.Right);

        // Visit current node last
        result.Add(node.Val);
    }

    // Start DFS from the root
    Dfs(root);

    // Return the collected results
    return result;
}

// ----------------------------
// Tree Printer (Sideways)
// ----------------------------

// Prints the tree "sideways" so structure is easy to see:
// - Right subtree appears on top
// - Left subtree appears on bottom
void PrintTree(TreeNode? root)
{
    Console.WriteLine("Tree (sideways):");
    PrintTreeImpl(root, 0);
    Console.WriteLine(); // blank line after the tree
}

// Number of spaces to indent per tree level
const int INDENT = 4;

// Recursive helper that prints the tree sideways
void PrintTreeImpl(TreeNode? node, int indent)
{
    // Base case: nothing to print
    if (node == null) return;

    // Print right subtree first so it appears "above" the current node
    PrintTreeImpl(node.Right, indent + INDENT);

    // Print the current node with indentation based on depth
    Console.WriteLine(new string(' ', indent) + node.Val);

    // Print left subtree last so it appears "below" the current node
    PrintTreeImpl(node.Left, indent + INDENT);
}

// ----------------------------
// Test Harness
// ----------------------------

// Build test tree:
//
//         1
//       /   \
//      2     3
//     / \   /
//    4   5 6
//
var root = new TreeNode(1)
{
    Left = new TreeNode(2)
    {
        Left = new TreeNode(4),
        Right = new TreeNode(5)
    },
    Right = new TreeNode(3)
    {
        Left = new TreeNode(6)
    }
};

// Print the tree visually
PrintTree(root);

// Run traversals (each returns a list of visited node values)
var pre = Preorder(root);
var inn = Inorder(root);
var post = Postorder(root);

// Print results as comma-separated lists
Console.WriteLine("Preorder:  " + string.Join(", ", pre));
Console.WriteLine("Inorder:   " + string.Join(", ", inn));
Console.WriteLine("Postorder: " + string.Join(", ", post));

// ==========================
// PREORDER TRAVERSAL (DFS)
// ==========================

// Returns the preorder traversal of a binary tree as a list of ints
// Order: Root -> Left -> Right
IList<int> Preorder(TreeNode? root)
{
    // This list will collect the traversal output
    var result = new List<int>();

    // Local recursive DFS function that performs the traversal
    void Dfs(TreeNode? node)
    {
        // Base case: empty subtree
        if (node == null) return;

        // Visit current node first
        result.Add(node.Val);

        // Visit left subtree
        Dfs(node.Left);

        // Visit right subtree
        Dfs(node.Right);
    }

    // Start DFS from the root
    Dfs(root);

    // Return the collected results
    return result;
}

// ==========================
// TREE NODE DEFINITION
// ==========================

// Simple binary tree node class used by the traversal methods
class TreeNode
{
    // Value stored in this node
    public int Val;

    // References to children (nullable because a node may not have children)
    public TreeNode? Left;
    public TreeNode? Right;

    // Constructor initializes the node with a value
    public TreeNode(int v) { Val = v; }
}

File: code_samples/section5/problems/problem5_1/problem5_1.py
"""Problem 5.1 - Traversals (Preorder, Inorder, Postorder)"""

from typing import List, Optional

# pylint: disable=too-few-public-methods


class TreeNode:
    """
    Model for a binary tree node.

    Each node stores:
    - val   : the integer value at this node
    - left  : reference to the left child (or None)
    - right : reference to the right child (or None)
    """
    def __init__(
        self,
        val: int = 0,
        left: "Optional[TreeNode]" = None,
        right: "Optional[TreeNode]" = None
    ):
        self.val = val
        self.left = left
        self.right = right


def preorder(root_node: Optional[TreeNode]) -> List[int]:
    """
    Preorder traversal (Root → Left → Right).

    Algorithm:
    1. Visit the current node
    2. Recursively traverse the left subtree
    3. Recursively traverse the right subtree
    """
    if root_node is None:
        return []

    return (
        [root_node.val]
        + preorder(root_node.left)
        + preorder(root_node.right)
    )


def inorder(root_node: Optional[TreeNode]) -> List[int]:
    """
    Inorder traversal (Left → Root → Right).

    Algorithm:
    1. Recursively traverse the left subtree
    2. Visit the current node
    3. Recursively traverse the right subtree

    In a Binary Search Tree (BST), this produces sorted order.
    """
    if root_node is None:
        return []

    return (
        inorder(root_node.left)
        + [root_node.val]
        + inorder(root_node.right)
    )


def postorder(root_node: Optional[TreeNode]) -> List[int]:
    """
    Postorder traversal (Left → Right → Root).

    Algorithm:
    1. Recursively traverse the left subtree
    2. Recursively traverse the right subtree
    3. Visit the current node

    Commonly used for deleting/freeing trees or evaluating expressions.
    """
    if root_node is None:
        return []

    return (
        postorder(root_node.left)
        + postorder(root_node.right)
        + [root_node.val]
    )


# ----------------------------
# Tree Printer (sideways)
# ----------------------------

def print_tree(root_node: Optional[TreeNode]):
    """
    Print the binary tree sideways for visualization.

    Right subtree is printed first so it appears at the top,
    left subtree is printed last so it appears at the bottom.
    """
    print("Tree (sideways):")
    _print_tree_impl(root_node, 0)
    print()


INDENT = 4  # number of spaces per tree level


def _print_tree_impl(node: Optional[TreeNode], indent: int):
    """
    Recursive helper for sideways tree printing.

    Parameters:
    - node   : current tree node
    - indent : horizontal spacing based on tree depth
    """
    if node is None:
        return

    # Print right subtree first (appears above)
    _print_tree_impl(node.right, indent + INDENT)

    # Print current node value with indentation
    print(" " * indent + str(node.val))

    # Print left subtree last (appears below)
    _print_tree_impl(node.left, indent + INDENT)


# ----------------------------
# Test Harness
# ----------------------------

# Build the test tree:
#
#         1
#       /   \
#      2     3
#     / \   /
#    4   5 6
#
root = TreeNode(
    1,
    left=TreeNode(
        2,
        left=TreeNode(4),
        right=TreeNode(5)
    ),
    right=TreeNode(
        3,
        left=TreeNode(6)
    )
)

# Print the tree visually
print_tree(root)

# Run traversals
pre = preorder(root)
inn = inorder(root)
post = postorder(root)

# Output results
print("Preorder: ", pre)
print("Inorder:  ", inn)
print("Postorder:", post)
